### PR TITLE
Enable use of add_subdirectory and build directory directly

### LINKED ISSUE #250: Incorporating Into An Existing CMake Project
What is the best way to incorporate pFUnit4 into an existing `CMake` project? 
It would be helpful to provide a guide similar to that [available in the googletest repository](https://github.com/google/googletest/tree/master/googletest#incorporating-into-an-existing-cmake-project), where this can be done via the `ExternalProject_Add()` `CMake` command.

### PR DESCRIPTION
**Pull request by ZedThree**
Alternative to #251 

Fixes #250 

This PR makes several changes that allow pFUnit to be consumed in a more consistent way through different methods:

- the currently supported way, via an installation
- using the build directory directly, via `-DCMAKE_PREFIX_PATH=/path/to/pfunit/build` or `-DPFUNIT_ROOT=/path/to/pfunit/build`
- as a git submodule, via `add_subdirectory(path/to/pfunit/submodule)`
- as an external project with `FetchContent` (or `ExternalProject`)

From my testing, this shouldn't affect existing uses at all, it just allows more uses.

Here's a little snippet that can be used to switch between downloading pfunit "on demand" (download at configure time, building at build time), and an existing build or install:

```cmake
if (DOWNLOAD_PFUNIT)
  message(STATUS "Downloading pFUnit...")
  include(FetchContent)
  FetchContent_Declare(
    pfunit
    GIT_REPOSITORY https://github.com/ZedThree/pFUnit
    GIT_TAG        export-cmake-targets
    )
  # How to set options for pFUnit
  set(SKIP_OPENMP YES CACHE BOOL "" FORCE)
  FetchContent_MakeAvailable(pfunit)
else()
  find_package(PFUNIT REQUIRED)
endif()

include(CTest)
enable_testing()
set(TEST_SOURCES
  tests/always_pass.pf)
add_pfunit_test(mytests
  "${TEST_SOURCES}" "" "")
target_link_libraries(mytests mylib)
```

This PR relies on some modifications to gFTL, gFTL-shared, and fArgParse
- https://github.com/Goddard-Fortran-Ecosystem/fArgParse/pull/74
- https://github.com/Goddard-Fortran-Ecosystem/gFTL/pull/112
- https://github.com/Goddard-Fortran-Ecosystem/gFTL-shared/pull/25

so it currently points the submodules to my forks/branches -- this will need adjusting once those PRs are merged.

A brief explanation of some of the changes:
- `CMAKE_SOURCE_DIR` is the top-level project's top-level directory, so this needs to be `PROJECT_SOURCE_DIR` in order to include pfunit via `add_subdirectory`
- targets need to be exported with `export(EXPORT ...)` and installed with `install(EXPORT ...)`
- removed the `external-pfunit` interface library as it was quite painful to get the targets exported correctly. It seemed to be used in only one place, so it didn't seem like much of an advantage to keep it
- use `CMakePackageConfigHelpers` module to write the package config and version files: this removes the need for the `PFUNITConfig-version.cmake.in` file
- put `add_pfunit_test` in a standalone file, so we can use it either from an installation or via `add_subdirectory`
- add alias targets with the `PFUNIT::` namespace to make sure linking can always be done against `PFUNIT::funit` for instance
- by default, don't build the pfunit tests when built as a subproject. `CTEST_CUSTOM_PRE_TEST` ends up clobbering the main project's tests otherwise

### CODE REVIEW

**Comment by ZedThree** [comment]
The tests are failing because macos-10.15/gfortran-8 failed to build MPI:

```
checking for C type matching bit representation of REAL*16... pending
./configure: line 60863:  5793 Abort trap: 6           ./conftest 1>&5 2>&1
checking if long double == REAL*16... Error!
configure: error: Could not determine if REAL*16 bit-matches C type
Error: Process completed with exit code 1.
```

**Comment by tclune** [comment]
Hmm.  That is just bizarre.   This part of the code has not changed in a very long time.

**Comment by tclune** [comment]
Hmm.  That is just bizarre.   This part of the code has not changed in a very long time.

Why is ./config even being invoked?

**Comment by ZedThree** [comment]
Sorry, to be clear, that snippet was from openmpi's `./configure`

**Comment by tclune** [comment]
Ah thanks.   Being barraged my email from many repos this morning.   I'll have my "CI Guy" look into this, but probably not until next week.

**Comment by Michael-A-Heuer** [comment]
Will it be included in the next pFUnit version and when will it be released approximately?

**Comment by tclune** [comment]
Sorry for the delays.   I had been hoping to have a person in my team look at the CI failures before too many more changes go in.    Last week got a bit crazy - seems that everyone tried to squeeze in 5 days work before the holiday.     Hopefully he can look at it today and clear things out.  Certainly by the weekend in the worst case.

I'll do a release as soon as everything is in place.  Lots of little things needed related to the submodules so it might take a few iterations.    I think 2 weeks is relatively safe, but hopefully  much sooner than that.

**Comment by Michael-A-Heuer** [comment]
Maybe you could just try to rerun the pipeline? Would be good to get @ZedThree's great contribution up and running.

**Comment by tclune** [comment]
Sure - I just activated that.   But I'm dubious.   I think there are fixes that are needed in the subrepos.

**Comment by Michael-A-Heuer** [comment]
It worked :)

**Review by tclune — APPROVED** [review]

File: .gitmodules [hunk 1]
```diff
@@ -1,10 +1,10 @@
 [submodule "gFTL"]
 	path = extern/gFTL
-	url = https://github.com/Goddard-Fortran-Ecosystem/gFTL.git
```
Comment (rafmudaf) [inline .gitmodules:3]: FYI these submodule links point to a different repo and it looks like this was meant to be changed before merged.

Comment (rafmudaf) [reply .gitmodules:3]: I'll submit a pull request to revert that, if you'd like.

Comment (ZedThree) [reply .gitmodules:3]: There are some PRs in those submodules that need to be merged first. But due to some circular dependencies, they were a little painful. They might actually be easier now this is merged

**Comment by tclune** [comment]
Yes - I did not have time to dig into the submodule issues this week, so figured we should get things approved that at least work. And if this helps to bootstrap all the better.      

While I still stand behind my reasons for incorporating gFTL into pFUnit,  the circularity is painful at times.   (gFTL uses pFUnit to test itself)

File: .gitmodules (hunk 1 above)
Comment (rafmudaf) [reply .gitmodules:3]: Ah ok thanks for the additional context.

**Pull request merged**